File: packages/python/pezhwan/tenant.py
"""Tenant management: CRUD operations for multi-tenant identities."""

from __future__ import annotations

from typing import Any, Mapping, Optional

from .client import PezhwanClient
from .errors import PezhwanApiError
from .models import Envelope


class Tenant:
    """Server-managed tenant resource."""

    def __init__(self, data: Mapping[str, Any]) -> None:
        self.id: str = str(data.get("id") or data.get("_id") or "")
        self.name: str = str(data.get("name") or "")
        self.slug: str = str(data.get("slug") or "")
        self.plan: str = str(data.get("plan") or "free")
        self.active: bool = bool(data.get("active", True))
        self.raw: Mapping[str, Any] = data


class TenantService:
    """Tenant lifecycle management via the identity server admin surface."""

    def __init__(self, client: PezhwanClient) -> None:
        self._client = client

    def create(
        self,
        name: str,
        slug: str,
        *,
        plan: str = "free",
        config: Optional[Mapping[str, Any]] = None,
    ) -> Tenant:
        """Create a new tenant."""
        body: dict[str, Any] = {"name": name, "slug": slug, "plan": plan}
        if config:
            body["config"] = config
        envelope = self._client.post("/v1/admin/tenants", body)
        return Tenant(envelope.data or {})

    def get(self, tenant_id: str) -> Tenant:
        """Fetch a single tenant by ID or slug."""
        envelope = self._client.get(f"/v1/admin/tenants/{tenant_id}")
        return Tenant(envelope.data or {})

    def list(self) -> list[Tenant]:
        """Return all tenants visible to the caller."""
        envelope = self._client.get("/v1/admin/tenants")
        items = envelope.data or []
        if isinstance(items, dict):
            items = items.get("tenants", [])
        return [Tenant(t) for t in items]

    def update(
        self,
        tenant_id: str,
        *,
        name: Optional[str] = None,
        plan: Optional[str] = None,
        active: Optional[bool] = None,
        config: Optional[Mapping[str, Any]] = None,
    ) -> Tenant:
        """Partially update a tenant."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if plan is not None:
            body["plan"] = plan
        if active is not None:
            body["active"] = active
        if config is not None:
            body["config"] = config
        envelope = self._client.post(f"/v1/admin/tenants/{tenant_id}", body)
        return Tenant(envelope.data or {})

File: packages/go/pezhwan/tenant.go
package pezhwan

import "context"

// Tenant describes a managed tenant resource.
type Tenant struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	Slug   string `json:"slug"`
	Plan   string `json:"plan"`
	Active bool   `json:"active"`
}

// TenantService manages tenant lifecycle via the identity server admin surface.
type TenantService struct {
	client *Client
}

// CreateTenant creates a new tenant.
func (s *TenantService) CreateTenant(ctx context.Context, name, slug string, opts ...TenantOption) (*Tenant, error) {
	body := map[string]any{"name": name, "slug": slug}
	for _, o := range opts {
		o(body)
	}
	data, err := s.client.post(ctx, "/v1/admin/tenants", body)
	if err != nil {
		return nil, err
	}
	return tenantFromMap(data), nil
}

// GetTenant fetches a tenant by ID or slug.
func (s *TenantService) GetTenant(ctx context.Context, id string) (*Tenant, error) {
	data, err := s.client.get(ctx, "/v1/admin/tenants/"+id)
	if err != nil {
		return nil, err
	}
	return tenantFromMap(data), nil
}

// ListTenants returns the tenants visible to the caller.
func (s *TenantService) ListTenants(ctx context.Context) ([]*Tenant, error) {
	data, err := s.client.get(ctx, "/v1/admin/tenants")
	if err != nil {
		return nil, err
	}
	items := stringifyList(data, "tenants")
	out := make([]*Tenant, 0, len(items))
	for _, m := range items {
		out = append(out, tenantFromMap(m))
	}
	return out, nil
}

// UpdateTenant partially updates a tenant.
func (s *TenantService) UpdateTenant(ctx context.Context, id string, opts ...TenantOption) (*Tenant, error) {
	body := map[string]any{}
	for _, o := range opts {
		o(body)
	}
	data, err := s.client.post(ctx, "/v1/admin/tenants/"+id, body)
	if err != nil {
		return nil, err
	}
	return tenantFromMap(data), nil
}

// TenantOption mutates a tenant request body.
type TenantOption func(map[string]any)

// WithTenantPlan sets the billing plan.
func WithTenantPlan(plan string) TenantOption {
	return func(b map[string]any) { b["plan"] = plan }
}

// WithTenantActive toggles tenant status.
func WithTenantActive(active bool) TenantOption {
	return func(b map[string]any) { b["active"] = active }
}

// WithTenantConfig sets the tenant configuration map.
func WithTenantConfig(config map[string]any) TenantOption {
	return func(b map[string]any) { b["config"] = config }
}

func tenantFromMap(m map[string]any) *Tenant {
	active, _ := m["active"].(bool)
	return &Tenant{
		ID:     stringValue(m["id"], m["_id"]),
		Name:   stringValue(m["name"]),
		Slug:   stringValue(m["slug"]),
		Plan:   stringValue(m["plan"]),
		Active: active,
	}
}

func stringValue(values ...any) string {
	for _, v := range values {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}